### DISCUSSION TITLE
Create a client instance in code running in docker container

### DISCUSSION THREAD

**Discussion by ScottLinnn** [kube-rs/kube#933] [Q&A] [answered]
Our code is using `kube::Client::try_default()` to initialize a client instance, and the code is running in a docker container. This is not working as I see in docker's logs: (kubeconfig: (failed to read kubeconfig from '"/root/.kube/config"': No such file or directory (os error 2))).   
One way I can think of to fix this is adding COPY in dockerfile to copy the config file into container, but due to our complicated docker-compose structure I couldn't do that right now. Are there other ways to get this work?  
Thanks!

**Comment by nightkr** [comment, marked as answer]
You typically don't want to bake the kubeconfig file into your image, since it will vary for each environment, and often contains some sort of credentials.

Instead, you have three other options:

1. Add a Docker volume that mounts ~/.kube into the container (see https://docs.docker.com/compose/gettingstarted/#step-5-edit-the-compose-file-to-add-a-bind-mount for how to do this)
    - This will probably not work (without additional research) if your cluster uses some external form of authentication (like GCP's OAuth, or a plugin), since those credentials will be stored elsewhere
2. Run it in Kubernetes instead, tools like https://tilt.dev/ help give a similar DX as docker-compose once set up
    - All Kubernetes containers automatically have a service credential mounted into them, so `Client:.try_default()` should "just work"
3. Run it outside of Docker
    - If your application only interacts with K8s through its API (as suggested by trying to access it from an out-of-cluster Docker container) then you might not need a container at all
    - Rust applications are usually statically(ish) linked, so many of the usual arguments for containerizing everything for environment consistency might not apply

**Reply by ScottLinnn** [reply]
Thank you so much!
